--- 
title: "SVG Tutorial for spatialHeatmap"
author: 
  Jianhai Zhang, PhD student at Genetics, Genomics and Bioinformatics, University of California, Riverside (<jzhan067@ucr.edu>; <zhang.jianhai@hotmail.com>). PI, Dr. Thomas Girke, professor at Department of Botany and Plant Sciences, University of California, Riverside (<thomas.girke@ucr.edu>)
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    fig_caption: true
    number_sections: true
    highlight: tango
    latex_engine: pdflatex
    keep_tex: true
  html_notebook: default
geometry: margin=1in
fontsize: 14pt
bibliography: bibtex.bib
package: spatialHeatmap
vignette: |
  %\VignetteIndexEntry{SVG tutorial for spatialHeatmap}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

  Maintainer: Jianhai Zhang (<jzhan067@ucr.edu>; <zhang.jianhai@hotmail.com>)

<style>body { text-align: justify }</style>

```{r setup0, echo=FALSE, message=FALSE, warning=FALSE}

library(knitr); opts_chunk$set(message=FALSE, warning=FALSE, tidy.opts=list(width.cutoff=65), tidy=TRUE)

```

# Summary

The R/Bioconductor package spatialHeatmap is designed for intuitive visualisation of large-scale data as long as a pair of configured data matrix and SVG image are provided. This tutorial is specifically devised for how to make an SVG image and configure it with the data matrix. 

To make a custom SVG image, a png image where regions should have clear contours, a data matrix, the SVG editor [Inkscape](https://inkscape.org/), and optionally the image editor [GIMP](https://www.gimp.org/) are required. The png image is the template while the data matrix is used to colour different regions in the SVG image. Inkscape is used to associate the SVG image with the data matrix and GIMP is optionally used to automatically extract polygons for the SVG image. 

In the following, the tutorial is given with a pair of configured gene expression matrix and SVG image of root tissues. All the files used in this tutorial can be downloaded [here](https://github.com/jianhaizhang/SVG_tutorial_file/tree/master/download) (download an SVG image: hover over the image, right click, and select "Save image as..."; download a PNG image: click the image, click "Download", right click, and select "Save image as..."; download a TXT file: click the file, click "Raw", right click, and select "Save as...".).

# Procedure

```{r, eval=TRUE, echo=FALSE, warnings=FALSE}

url <- "https://raw.githubusercontent.com/jianhaizhang/SVG_tutorial_file/master/image/"

```
## Gene Expression Matrix

The gene expression matrix usually contains noise and needs to be filtered at the beginning. The function `filter.data` [@SummarizedExperiment; @data.table; @utils] has two filter arguments `pOA` and `CV`, corresponding to `pOverA` and `cv` in the package "genefilter" [@Gentleman2018-xj], respectively.   

<a name="name_format"></a>
In the gene expression matrix, the row and column names should be gene IDs and sample/conditions respectively. The sample/condition names MUST be fomatted this way: a sample name is followed by double underscore then the condition, such as â€œepidermis__140mM_1h" in [Table 1](#table) [@Geng2013-ek], where epidermis is the sample and 140mM_1h is the condition. In the column names of sample/condition, only letters, digits, single underscore, single space, or dots are allowed. Not all samples in the matrix necessarily need to be present in the SVG image, vice versa. Only samples present in the SVG image are recognised and coloured.

The expression matrix is stored as an "SummarizedExperiment" object. Metadata of genes and sample/conditions can be optionally added. Refer to the R package "SummarizedExperiment" for more details [@SummarizedExperiment].

```{r, filter, eval=TRUE, echo=TRUE, warnings=FALSE }

  library(spatialHeatmap); library(data.table); library(SummarizedExperiment)
  # Creat the "SummarizedExperiment" class.
  ## The expression matrix, where the row and column names should be gene IDs and sample/conditions, respectively.
  data.path <- system.file("extdata/example", "root_expr_row_gen.txt", package = "spatialHeatmap")
  expr <- fread(data.path, sep='\t', header=TRUE, fill=TRUE)
  col.na <- colnames(expr)[-ncol(expr)]; row.na <- as.data.frame(expr[, 1])[, 1]
  expr <- as.matrix(as.data.frame(expr, stringsAsFactors=FALSE)[, -1])
  rownames(expr) <- row.na; colnames(expr) <- col.na
  con.path <- system.file("extdata/example", "root_con.txt", package = "spatialHeatmap") 
  ## Condition is a single column data frame.
  con <- read.table(con.path, header=TRUE, row.names=NULL, sep='\t', stringsAsFactors=FALSE)
  ann.path <- system.file("extdata/example", "root_ann.txt", package = "spatialHeatmap")
  ## Gene annotation is a single column data frame.
  ann <- read.table(ann.path, header=TRUE, row.names=1, sep='\t', stringsAsFactors=FALSE)
  ## The expression matrix, gene annotation, and condition are stored in a "SummarizedExperiment" object. Gene annotation and condition are optional.
  expr <- SummarizedExperiment(assays=list(expr=expr), rowData=ann, colData=con)

  # Filter genes. In "pOA", genes with expression value A >= 1 in at least p=0.03 (3%) of all samples are retained; in "CV", genes with coefficient of variance (cv) between 0.1 and 10000 are retained, where the upper limit is set to very high (10000) so as to keep all genes with cv over 0.1. 
  exp <- filter.data(data=expr, pOA=c(1, 0.03), CV=c(0.1, 10000), dir=NULL)

  # Get the filtered matrix. "filter.data" returns a "SummarizedExperiment" object.
  df <- assay(exp)

```
<a name="table"></a>
```{r, table, eval=TRUE, echo=FALSE, warnings=FALSE}

  kable(df[1:3, 1:4], caption='Table 1. Gene expression matrix. Rows and columns are genes and sample/conditions respectively.')

``` 

## Make SVG Images 

### <a name="draw"></a> Make Blank SVG Images by Drawing 

If the contour in the png image is not clear, GIMP can generate low-quality SVG images, so in this case one can draw the blank SVG image with Inkscape by using the png as a template. Below is an example of drawing only two polygons.

  1. Draw polygons <p/><p/> 
  Open the root png image [@Mustroph2009-nu] in Inkscape. The image can be zoomed by press "-" or "+" on the keyboard. Select the "Draw freeahnd lines (F6)" at the left tool bar. Left click once at the first corner of the polygon, move to the second corner and double left click, and so on. Lastly, after drawing the last line click at the first corner to seal the polygon.

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/open.png", "img/draw_freehand_line.png", "img/double_click.png", "img/draw_second_line.png", "img/draw_last_line.png"))

```

  2. Align polygons <p/><p/> 
  Select "Edit path by nodes (F2)" from the left tool bar. Draw a large rectangle to select the whole sealed polygon and draw another large rectangle to select all nodes.

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/edit_path_tool.png", "img/select_path.png", "img/select_path1.png", "img/select_node.png", "img/select_node1.png"))

```

    Click the "Make selected nodes smooth" in the top tool bar. Drag the edges and handles to make the sealed polygon aligned with the template, then the first polygon is finished. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/make_selected_nodes_smooth.png", "img/drag_handle.png", "img/first_polygon.png"))

```

  3. <a name="shape"></a> Make polygons with shapes <p/><p/>
  Alternatively, the polygons can be made with rectangles. Click "Create rectangles and squares (F4)" at the left tool bar and draw a rectangle in the second template polygon. Select the rectangle and click "Object to Path" under the "Path" tab at the top, then the rectangle becomes an SVG path, which can be edited. Switch cursor to "Select and Transform Objects (F1)" and rotate the rectangle as expected. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/create_rectangle.png", "img/draw_rectangle1.png", "img/object_to_path1.png", "img/select_tool.png", "img/rotate_rectangle.png"))

```

  &nbsp; &nbsp; Switch the cursor to "Edit path by nodes (F2)" and select all nodes of the rectangle. Click "Make selected nodes corner" at the top tool bar. Drag the edges and handles to overlay the rectangle path on the template, then the second polygon is done. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/select_rectangle_node.png", "img/make_node_corner.png", "img/drag_rectangle.png", "img/second_polygon.png"))

```

  In this root image there are many polygons, so it takes too much time to draw them individually. However, GIMP can be used to extract the polygons automatically. <p/>

### Make Blank SVG Images with GIMP

If the png image contains many polygons and their contours are clear (e.g. the root image), GIMP can be used to automatically extract the polygons. <p/>

  4. Open the root png image with GIMP. There are five coloured tissues (skyblue, green, red, yellow, blue). The skyblue is not present in the gene matrix while the green, red, yellow, blue correspond to epidermis, cortex, endodermis, stele in the gene matrix respectively. Enable the "Paths" panel under the "Windows" tab at the top. Right click and select "By Color" under "Select". <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/open1.png", "img/path_panel.png", "img/by_color.png"))

```

  5. Click on the skyblue tissue, then right click, select "To Path" under "Select", and the paths of all skyblue tissues are extracted in the "Paths" panel. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/click_first_color.png", "img/to_path.png", "img/first_path.png"))

```

  6. Similarly, exrtact paths of other tissues. In the "Paths" panel set paths visible by enabling the eye symbol, then right click and merge visible paths. Right click and export merged paths, which is a blank SVG image. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/eye.png", "img/merge.png", "img/export.png"))

```

### Make Blank SVG Images with GIMP and Drawing

The drawing method creates accurate SVG images but it is time-consuming, while the GIMP method is faster but it can generate fused polygons. In this tutorial, the root png image has well-separated polygons and clear contours, so GIMP can produce accurate SVG images. Otherwise the resulting SVG images would have mixed and noise polygons. Considering the pros and cons of two methods, the good practice is to first use GIMP to extract polygons then use drawing method to refine them. Below is an example of an SVG image with fused polygons, which is generated by GIMP.

  7. <a name="layer"></a> Place all SVG paths inside a layer <p/><p/>
  Open the blank fused SVG image in Inkscape (the image can be zoomed by press "-" or "+" on the keyboard). Under the "Layers" tab at the top click "Layers", the "Layers" panel will come out on the right, then click "+" at the bottom left corner of the panel to add "Layer 1". 

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/layer1.png", "img/layer2.png", "img/add_layer.png"))

```

    Draw a rectangle over the fused SVG graph and cut. Then click on "Layer 1" and paste the fused SVG into "Layer 1" (make sure the "Layer 1" is unlocked by refering to the lock symbol). Open the "XML Editor" from the "Edit" tab at the top, if "\<svg:path id="path 77"\>" is under "\<svg:g id="layer1" inkscape:label="Layer 1"\>", then the fused SVG is inside \"Layer 1\".  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/draw_rectangle.png", "img/click_layer.png", "img/paste_layer.png", "img/to_layer.png", "img/in_layer.png"))

```

  8. Refine and modify the SVG image <p/><p/>
  Draw a ractangle over the image, click "Break Apart" under the "Path" tab, then select the outer noisy rectangle by clicking on its edge. Press "delete" on the keyboard to delete it. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/break_apart1.png", "img/remove_outer_frame.png"))

```

    Click the edge of the large fused polygon and move. Use the [drawing method](#draw) to make new polygons (blue) with the fused ones as templates. Delete the fused polygons and move back the new polygons to make the tissue complete. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/move_path.png", "img/aligned_3_polygons.png", "img/new_polygon.png"))

```

If a large fused polygon needs to be separated, one can use the eraser tool. Drag the fused polygon away from the tissue, select the eraser tool from the left tool bar, then use the eraser to cut the fused polygon into three independent polygons. Select the cut polygons and click "Break Apart" under the "Path" tab at the top, then the three polygons are separated. Place back the three polygons to make the tissue complete.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/linked_polygon.png", "img/eraser.png", "img/erase1.png", "img/erase3.png"))

```

  9. Fill and stroke <p/><p/>
  Under "Object" tab at the top, select "Fill and Stroke...", then the "Fill and Stroke (Shift+Ctrl+F)" panel will come out on the right. Select all polygons by drawing a large rectangle over them. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/fill_tab.png", "img/select_all1.png"))

```

    <a name="opacity"></a> Under the "Stroke paint" tab in the fill and stroke panel, select "Flat color". Under the "Stroke style" tab, set the stroke width, e.g.: 1.5 px. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/stroke_flat_color.png", "img/stroke_width.png"))

```

    Under the "Fill" tab, click "No paint" to get a blank SVG image, which is ready to use in next section. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/fill_no_paint.png"))

```

## Configure the SVG Image with Gene Matrix

  In the SVG image, each polygon has a unique ID. To plot spatial heatmaps, these IDs should be exactly replaced with sample names in the gene matrix. No matter how the blank SVG image is created, it should be [placed inside a layer](#layer) before start the following steps.

  10. Group same tissues <p/><p/>
  If multiple polygons belong to the same tissue type, they should be grouped together. The example of grouping epidermis is given below. Open "XML Editor..." under the "Edit" tab at the top, then the "XML Editor (Shift+Ctrl+X)" panel comes out on the right. Click all the epidermis polygons while pressing the "Shift" key, right click, and select "Group". A group should not contain another group.

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/select_group.png", "img/group.png"))

```

  11. Associate polygons with samples <p/><p/>
  The epidermis group \<svg:g id="987"\> shows up in the XML panel. Click "id", change "g987" to "epidermis", and click "Set", then the new group id is set.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/click_id.png", "img/set_id.png" ))

```

In the "Fill and Stroke (Shift+Ctrl+F)" panel, select "Flat color" under the "Fill" tag, then specify a color for epidermis.      

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics("img/flat_color.png")

```

Group other tissues, set ids and colours.   

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics("img/colored.png")

```

  12. Polygon orders <p/><p/>
  The polygons are stacked over each other according to their orders in the "XML Editor", so the first polygon might be invisible because it could be covered by the second, third, and so on. For instance, in the brain SVG image [@brain; @hipp], the grey outline polygon (the path "rect5480") is the first one, and partially covered by other polygons. Therefore, users should drag and organise the paths in expected order. <p/>

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/brain1.png", "img/brain2.png"))

```

## Add Text to Label Tissues

  Users can add text to label tissues. Basically, the text is first typed in with the text tool and then the text object is coverted to paths. Next paths are added into the polygon group of target tissue and filled with the same tissue colour. Below is the example of adding text to the epidermis tissue.

  13. Creat text paths <p/><p/>
  Select "Creat and select text objects (F8)" from the left tool bar, drag a text box, and type epidermis. Click on the text object and convert it to path. 

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/creat_text.png", "img/text_box.png", "img/type.png", "img/object_to_path1.png"))

```

  14. Text fill and stroke <p/><p/>
  Click on the text paths and fill them with the same colour of epidermis using "Pick colors from image (F7)" from the left tool bar. In the "Fill and Stroke" panel, set the stroke style.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/pick_color_tool.png", "img/epidermis_green.png"))

```

  15. Add text paths to tissue group <p/><p/>
  Click and cut the text, then double click epidermis (green) to enter the group.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/enter_group.png"))

```

    Paste the text anywhere then the text is inside the epidermis group as a "text group". Move/resize the text group as expected. Right click and ungroup the text.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/ungroup_text.png"))

```

    All the letters are in the epidermis group as individual paths, which can be seen in the XML Editor.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/ungroup_text1.png"))

```

  To add a pointer, [draw a rectangle](#shape) and convert it to path, fill it with the same style as epidermis. Move, rotate and resize it as expected.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/drag_pointer.png"))

```

  Clicking any of other tissues will group the text, pointer, and epidermis together, which can be confirmed by dragging epidermis to see they are moving as a whole.  

```{r eval=TRUE, echo=FALSE, fig.align="center", fig.cap=NULL, out.width="60%", fig.show='hold' }

include_graphics(c("img/drag_new_group.png"))

```

  Similarly, add text to label other tissues.  

  17. Save the final SVG image, which is ready to use in the spatialHeatmap. The name of saved SVG image can only consists of letters, digits, underscores. E.g. "root_cross_final.svg" is acceptable while "root_cross_final(copy).svg" will give errors. <p/>

## Troubleshooting

  1. Make sure the coloumn names are formatted in the right way.
  2. Make sure the target tissue names are exactly same between data matrix and SVG.
  2. Make sure a layer (or group) contains all the paths, groups in the XML editor.
  3. Make sure a group does not contain any other groups.

# Reference


