<!DOCTYPE html>
<html>
<body>

<div align="justify">
<p>For a quick test, select one example under "Select a work mode", or select "Compute online" and download the aSVG (see below) image and data matrix from below. </p>

<p>This Shiny app is an integrated implementation of the R/Bioconductor package "spatialHeatmap". It is designed for interactively visualizing numeric values in biological assays (e.g. RNA-seq, microarray, qPCR, etc.) on an anotomical aSVG image. First, the core feature "Spatial Heatmap" maps numeric values of target assayed items (genes, proteins, metabolites, etc) to matching spatial features (cells, tissues, organs, etc.) in the aSVG image. In this process, the values are translated into different colors and these colors are used to paint the spatial features. The resulting images are called spatial heatmaps. Second, nearest neighbors of each target assayed item are selected by most similar expression profiles independently. All target items and their nearest neighbors are hierachically clustered and visualized as interactive matrix heatmap. Third, network modules are identified internally and the module containing a target item is displayed as interactive network graph. The network module identification is computationally intensive for large data matrix (e.g. > 10,000 rows). Therefore, the "Compute locally" mode is developed accordingly. If the data matrix is small (e.g. < 10,000 rows) it can be uploaded in the "Compute online" mode. In the following, instructions are given with a gene expression matrix. However, this app is a general visualization tool, not limited to biological data.  </p>

<p>When switch among different modes, users should refresh the webpage before selecting the new mode. Operations on this app are expected to follow the order as each panel appears. Otherwise, errors may arise. If so, the solution is to refresh the webpage.</p>
</div>


</body>
</html>







